### PR TITLE
Fix confusing `default=True` bool behavior (Fix #68)

### LINKED ISSUE #68: Confusing command line options for bool's with `default=True`
Having a boolean option with a default value of `True`:
```
from dataclasses import dataclass

from simple_parsing import ArgumentParser


@dataclass
class Args:
    enable_water: bool = True  # Do you want water?


parser = ArgumentParser()
parser.add_arguments(Args, dest="args")

print(parser.parse_args(["--enable_water"]))
# => Namespace(args=Args(enable_water=False))
```

Creates this very confusing behaviour that passing the option (without a boolean value) disables it, so passing `enable_water` actually causes `enable_water` to be set to False, which is not very intuitive.

Even though this behaviour corresponds to `store_false` in the classic python `ArgumentParser`, it would be nicer to make this more clear. For example by changing the command line option automatically to `--no_enable_water` (by prepending a `no_`) when the default is `False`, and then setting `enable_water` to False when the option `--no_enable_water` is passed.

### PR DESCRIPTION
**Pull request by Conchylicultor**
I implemented a custom `argparse.Action`. Indeed, this is the **official** solution of argparse added in 3.9: [`argparse.BooleanOptionalAction`](https://docs.python.org/3/library/argparse.html#:~:text=argparse.booleanoptionalaction)

This implementation is inspired from the [official argparse implementation](https://github.com/python/cpython/blob/72263f2a20002ceff443e3a231c713f2e14fe3fe/Lib/argparse.py#L878) with some improvements:

* Support custom `--no` prefix instead of `--no-` (for `absl.flags` compatibility).
* Accept `--flag=true` value
* Support `nargs`
* Support Python 3.8

Mutually exculsive group would be:

* More hacky (flags are parsed independently, so it's difficult to have a global logic)
* Flags would appear either twice int the doc or not be documented
* `argparse.Action` has better error messages, like: `error: the following arguments are required: --my_flag/--nomy_flag`
* Inconsistent with how args are resolved (for example, it is sometimes convenient to be able to overwrite a flag value, for example in bash aliases: `alias my_prog="my_prog --dry_run"`)

### CODE REVIEW

File: simple_parsing/wrappers/field_wrapper.py
Comment (lebrice) [inline simple_parsing/wrappers/field_wrapper.py:1102]: Just to make sure I totally understand: This means you can set `--verbose=True` but can't set `--noverbose=False`?

File: simple_parsing/wrappers/field_wrapper.py
Comment (lebrice) [inline simple_parsing/wrappers/field_wrapper.py:1113]: Would this usage str also be correct when using `nargs`?

File: simple_parsing/wrappers/field_wrapper.py
Comment (lebrice) [inline simple_parsing/wrappers/field_wrapper.py:1097]: Shouldn't `self.default` be used at some point? Or is the value ignored? What happens if the flag isn't actually passed?

**Review by lebrice — CHANGES_REQUESTED** [review]
Thanks a lot @Conchylicultor, this is very helpful.

I left a bunch of nitpicky suggestions / comments.
I believe there are a few more things we could explore before merging this.

- How does someone customize the prefix?
It should be easier than setting the module-global variable. It should also ideally be customizable on a per-field basis, rather than one global setting.

I have some suggestions to help with this. For instance, there is already a `flag` utility function for defining a field. Perhaps this function could accept a `negative_prefix="--no"` argument, that would then get stored in the field metadata (the `custom_argparse_args` entry, IIRC).
The `_arg_options` dict is updated with the contents of this dict, so the values would end up getting passed as constructor arguments to your custom Action class later when `add_arguments(**field_wrapper.arg_options)` happens.
There should also be a minimal test to check that this works.

- Will there be backward incompatibility issues? If so, how do we deal with them?
I'm not sure what the right thing to do is here, but it's clear that a small test/example that illustrates the difference would be useful.

Let me know what you think. I know this may seem like a lot, but we can move some of the stuff I mentioned to another PR as well if you want.
 
Again, thank you very much for making this PR. I really appreciate it.

File: simple_parsing/wrappers/field_wrapper.py
Comment (lebrice) [reply simple_parsing/wrappers/field_wrapper.py:1059]: Oh, one thing might be important to check with this. Do we set a value of `nargs=None` in the _arg_options dict?
If so, we'd need to not set the entry in the first place.

File: simple_parsing/wrappers/field_wrapper.py
Comment (Conchylicultor) [reply simple_parsing/wrappers/field_wrapper.py:1113]: `--noflag` is only added when `nargs == '*' '?'` or `nargs >= 1`, so it won't appear in the help otherwise.

File: test/test_bools.py
Comment (Conchylicultor) [reply test/test_bools.py:88]: I find having the `pytest.mark.parametrize` to have the 4 same examples for each `nargs=` makes it more readable easier to think about the difference between each `nargs=`.

**Review by Conchylicultor — COMMENTED** [review]
* Customizing the `--no` prefix:

Indeed currently this is not possible. Because this seems a quite independent feature. I think it make sense to add it in a separate PR.

* Breaking compatibility:

This will change the default behavior when the default is `True`, which is the current #68 bug this PR fix. With `verbose: bool = True`:

* Before: `--verbose` would set `args.verbose = False` (bug)
* After: `--verbose` set `args.verbose = True`

File: simple_parsing/wrappers/field_wrapper.py
Comment (lebrice) [reply simple_parsing/wrappers/field_wrapper.py:1048]: I see what you mean, but I meant having them be annotated directly. 
Also, correct me if I'm wrong, but IDEs wouldn't look for the type hints in the argparse typeshed when using simple-parsing, right?

File: simple_parsing/wrappers/field_wrapper.py
Comment (lebrice) [reply simple_parsing/wrappers/field_wrapper.py:1056]: Sure, then I suggest using `int | Literal["?", "*", "+"] | None`

File: simple_parsing/wrappers/field_wrapper.py
Comment (lebrice) [reply simple_parsing/wrappers/field_wrapper.py:1059]: Ah I see, agreed.

File: simple_parsing/wrappers/field_wrapper.py
Comment (lebrice) [reply simple_parsing/wrappers/field_wrapper.py:1065]: I don't see the changes though. Did you push your commit? or am I viewing the wrong version?

File: simple_parsing/wrappers/field_wrapper.py
Comment (lebrice) [reply simple_parsing/wrappers/field_wrapper.py:1097]: Ok I'll take a look at the tests and resolve this after.

File: simple_parsing/wrappers/field_wrapper.py
Comment (lebrice) [reply simple_parsing/wrappers/field_wrapper.py:1102]: I could see myself agreeing with this idea, but as-is, it doesn't seem to be compatible with having nargs="*", right?

I think even though it's confusing, it might be best to keep things uniform for now, and have `--noverbose=False` be the same as `--verbose=True`. We can then discuss this in an issue and turn this off later if needed.

Comment (lebrice) [reply simple_parsing/wrappers/field_wrapper.py:1102]: Ah nevermind about nargs="*", you only create the negative flag when nargs is for a single argument.
I agree. Let's leave it like this for now then.

File: simple_parsing/wrappers/field_wrapper.py
Comment (lebrice) [reply simple_parsing/wrappers/field_wrapper.py:1113]: I see. Makes sense.

File: test/test_bools.py
Comment (lebrice) [reply test/test_bools.py:88]: I see your point, but I still consider it important to test the error messages that we get. 
The `match` argument of `pytest.raises` doesn't work perfectly in this case, so you could use `testutils.exits_and_writes_to_stderr`, for instance:
```python
with exits_and_writes_to_stderr("--no cannot be used with value"):
    _ = Foo.setup("--nodebug=True")
```

**Review by lebrice — APPROVED** [review]
Hi @Conchylicultor, thanks again for this.

Added some clarifications. I'd appreciate the changes but if you're not up to it, let me know, no worries, I can also add the changes myself in another PR.

**Comment by Conchylicultor** [comment]
Oups, sorry, it looks like VS code didn't pushed my changes.
I'll fix this tomorrow.

(Sorry, I'm not really used to the open source workflow anymore)

File: simple_parsing/wrappers/field_wrapper.py
Comment (Conchylicultor) [reply simple_parsing/wrappers/field_wrapper.py]: Pushed my changes. Not that `str` is better than `Literal` here because there's some undocumented argparse value for `nargs`. Currently, those will raise error.

File: simple_parsing/wrappers/field_wrapper.py
Comment (Conchylicultor) [reply simple_parsing/wrappers/field_wrapper.py]: Sorry about that. Pushed the changes

File: test/test_bools.py
Comment (Conchylicultor) [reply test/test_bools.py]: Updated the tests to check the error message

File: simple_parsing/wrappers/field_wrapper.py
Comment (Conchylicultor) [reply simple_parsing/wrappers/field_wrapper.py]: Done

**Comment by Conchylicultor** [comment]
I pushed my changed. Sorry for the noise.

Let me know what you think

File: simple_parsing/wrappers/field_wrapper.py
Comment (lebrice) [reply simple_parsing/wrappers/field_wrapper.py]: What value? In which case would it raise an error?

File: test/test_bools.py
Comment (lebrice) [inline test/test_bools.py:96]:
```suggestion
        """Some test class"""
```

File: test/test_bools.py
Comment (lebrice) [inline test/test_bools.py:97]: By setting `nargs` explicitly, you're bypassing the logic that generates the `nargs` from the annotation and context.
This might be fine, if you're just testing the new Action class on it's own. 
The only thing that bugs me is that the field annotation here is often incorrect (here with nargs="*" you should have a `list[bool]` annotation instead of `bool`).
It could be fine in some situations to manually specify `nargs`, for example if you have a `list[bool]` annotation, you could pass `nargs="+"` or `nargs > 1` to specify the number of values you expect.
Also, FYI, when using `tuple[bool, bool, bool]`, the generated `nargs=3`.

**Review by lebrice — APPROVED** [review]
Thanks!

**Comment by lebrice** [comment]
Ok I'm happy with this. Thanks again @Conchylicultor. I'll update the branch and merge it.
I don't think backward-compatibility will be that big of an issue on this. Hopefully this doesn't come back to bite me later!

File: test/test_bools.py
Comment (Conchylicultor) [reply test/test_bools.py:97]: > By setting nargs explicitly, you're bypassing the logic that generates the nargs from the annotation and context.

But this is a limitation of SimpleParsing right ? For example:

```python
@dataclass
class MyClass(TestSetup):
        a: int = helpers.field(nargs='*')
    
flags = MyClass.setup('--a 1 2 3')
assert flags.a == [1, 2, 3]
```

So I don't think this is related to this PR.

**Comment by Conchylicultor** [comment]
Fixed the test

> Does that change make sense?

The `[<>]` is a convention to specify that the `<>` is optional (so `--a [bool]` match `--a` or `--a True`).
Ideally, this should depend on `nargs`:

* `'?'`: `--a [bool]`
* `1`: `--a bool` 
* `'+'`: `--a bool [bool ...]`

But it's likely overengineered and I don't think this should block this PR

File: test/test_bools.py
Comment (lebrice) [reply test/test_bools.py:97]: It's not exactly a limitation of `simple-parsing`. it's just that the annotation you're using (`int`) becomes incorrect if you set `nargs="*"` explicitly.

If you had this, then it would be correct:
```python
@dataclass
class MyClass(TestSetup):
        a: list[int]

flags = MyClass.setup('--a 1 2 3')
assert flags.a == [1, 2, 3]
```

File: simple_parsing/wrappers/field_wrapper.py
Comment (lebrice) [inline simple_parsing/wrappers/field_wrapper.py:1141]: Yeah that's a good point, although since this action is only used with plain `bool` fields, then it should probably be fine for now. This action isn't used when you have a `bool | None` field, or `list[bool]`, etc, so the `nargs` would only be different if it were set manually with the `field` function.

I guess we could fix this very minor issue later.

**Review by lebrice — APPROVED** [review]

File: simple_parsing/wrappers/field_wrapper.py
Comment (Conchylicultor) [reply simple_parsing/wrappers/field_wrapper.py]: I'm talking of `argparse.REMAINDER`, `argparse.PARSER`...: https://github.com/python/cpython/blob/cb6ff54b01118e01c26758d2f57cf6da60277bd4/Lib/argparse.py#L101

No idea what those are doing. Currently this will raise an explicit error message in:

`if not (isinstance(nargs, int) or nargs in {'?', '*', '+'}):` would raise an error.

I think this is fine for now. If really required, we could add support for those later.

File: test/test_bools.py
Comment (Conchylicultor) [reply test/test_bools.py:97]: I just meant if we want to fix this behavior (e.g. prevent `nargs='+'` for non-list types), this should be done in another PR as this behavior is not related to this specific action.

File: simple_parsing/wrappers/field_wrapper.py
Comment (Conchylicultor) [reply simple_parsing/wrappers/field_wrapper.py:1141]: Actually I think the `[]` should be automatically added here: https://github.com/python/cpython/blob/cb6ff54b01118e01c26758d2f57cf6da60277bd4/Lib/argparse.py#L595

Not sure why this isn't the case. Because this is such a minor issue. I'll be in favor of merging this now and investigate this later.

**Comment by lebrice** [comment]
Hey @Conchylicultor would you mind just correcting the expected strings that are causing the tests to fail?
I'll merge the PR once those pass.

File: simple_parsing/wrappers/field_wrapper.py
Comment (lebrice) [reply simple_parsing/wrappers/field_wrapper.py:1141]: Sure, fair enough

File: test/test_bools.py
Comment (lebrice) [reply test/test_bools.py:97]: Fair enough!

**Comment by Conchylicultor** [comment]
Pushed a fix. Tests pass locally for me.

**Pull request merged**
